### PR TITLE
Modified LoginServlet for Password functionality

### PR DESCRIPTION
**Pull request by trishazalani**
I have modified LoginServlet for Password functionality, test cases will be modified in the next commit. Please let me know if there are any changes to be made.

### CODE REVIEW

File: src/main/java/codeu/controller/LoginServlet.java [hunk 1]
```diff
@@ -59,28 +59,34 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
   }
 
   /**
-   * This function fires when a user submits the login form. It gets the username from the submitted
-   * form data, and then adds it to the session so we know the user is logged in.
+   * This function fires when a user submits the login form. It gets the username and password from the submitted
+   * form data, checks that they're valid, and either adds the user to the session so we know the user is logged in.
```
Comment (nl356) [inline src/main/java/codeu/controller/LoginServlet.java:62]: Extraneous period

**Comment by TylerSandman** [comment]
In general, it's good to split up commits into smaller chunks like you're doing here (functionality in this commit, and tests in another). So good thinking!

For reference, it's more important that for every pull request, all tests still pass. We should never merge pull requests where tests are failing even if it means making our PR a little larger. I trust that whoever is reviewing will also double check that tests pass. That is, we should wait for the commit with the tests (in the same PR before merging).

File: src/main/java/codeu/controller/LoginServlet.java (hunk 1 above)
Comment (trishazalani) [reply src/main/java/codeu/controller/LoginServlet.java:62]: I'm unsure of what you meant, could you clarify?

**Comment by trishazalani** [comment]
Should I undo this request, or let it stay and update the test cases in another?

**Comment by TylerSandman** [comment]
I should have clarified more; sorry. This is fine as long as the tests are added as a separate commit to this same PR before merging. The comment above was more of an FYI.

File: src/main/java/codeu/controller/LoginServlet.java (hunk 1 above)
Comment (nl356) [reply src/main/java/codeu/controller/LoginServlet.java:62]: Just a minor typo in your commenting "user is logged in. or shows an error"

Comment (trishazalani) [reply src/main/java/codeu/controller/LoginServlet.java:62]: Oh okay. Got it!

**Comment by trishazalani** [comment]
Alright. I'll just add the tests as another commit to this branch once I'm dne

**Comment by nl356** [comment]
Updated Test Cases aren't passing for me... seems like a null pointer exception somewhere
screenshot from 2018-04-03 18-28-53

**Comment by trishazalani** [comment]
Thats strange. My branch seems to be passing all. Are you sure you are using the latest commit?

**Comment by nl356** [comment]
Oh, I think this may be because you made two different branches? "tzalani" and "Project1_tzalani".
I believe you're updating for the same functionality of LoginServlet and consequently its tests, so could you just commit that change here? I think putting them separately made me and Han confused lol.

File: src/test/java/codeu/controller/LoginServletTest.java [hunk 2]
```diff
@@ -77,21 +78,32 @@ public void testDoPost_NewUser() throws IOException, ServletException {
...
     loginServlet.doPost(mockRequest, mockResponse);
 
-    ArgumentCaptor<User> userArgumentCaptor = ArgumentCaptor.forClass(User.class);
+    Mockito.verify(mockRequest)
+            .setAttribute("error", "That username was not found.");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
 
-    Mockito.verify(mockUserStore).addUser(userArgumentCaptor.capture());
-    Assert.assertEquals(userArgumentCaptor.getValue().getName(), "test username");
+    //ArgumentCaptor<User> userArgumentCaptor = ArgumentCaptor.forClass(User.class);
```
Comment (TylerSandman) [inline src/test/java/codeu/controller/LoginServletTest.java]: Here and below: we should never commit commented out code. Prefer removing it altogether please.

File: src/test/java/codeu/controller/LoginServletTest.java [hunk 3]
```diff
@@ -104,4 +116,31 @@ public void testDoPost_ExistingUser() throws IOException, ServletException {
...
+    loginServlet.setUserStore(mockUserStore);
+
+    HttpSession mockSession = Mockito.mock(HttpSession.class);
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
+
+    loginServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verify(mockUserStore, Mockito.never()).addUser(Mockito.any(User.class));
+
+    Mockito.verify(mockRequest)
```
Comment (TylerSandman) [inline src/test/java/codeu/controller/LoginServletTest.java]: Nit: please keep indentation consistent or put this all on one line. Everywhere else in the code as far as I can tell uses 4 spaces for line wrapping indentation ([style guide for reference](https://google.github.io/styleguide/javaguide.html#s4.5.2-line-wrapping-indent))

Same thing for line 82

**Comment by trishazalani** [comment]
Okay. I will re-try pushing the commits again.

**Comment by trishazalani** [comment]
So, I've checked and everything says up to date. Are you checking the latest commit?

**Review by TylerSandman — CHANGES_REQUESTED** [review]
LGTM with some nits. Please get another person to LGTM and make sure all the tests are passing.

**Review by TylerSandman — APPROVED** [review]

File: src/main/java/codeu/controller/LoginServlet.java [hunk 4]
```diff
@@ -59,28 +58,33 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
-    if (!username.matches("[\\w*\\s*]*")) {
-      request.setAttribute("error", "Please enter only letters, numbers, and spaces.");
-      request.getRequestDispatcher("/WEB-INF/view/login.jsp").forward(request, response);
-      return;
+    if (userStore.isUserRegistered(username)) {
+      User user = userStore.getUser(username);
+      if(password.equals(user.getPassword())){
+        request.getSession().setAttribute("user", username);
+	    response.sendRedirect("/conversations");
+      }else{
```
Comment (TylerSandman) [inline src/main/java/codeu/controller/LoginServlet.java]: Nit: Add spaces before and after else

}else{ --> } else {

File: src/main/java/codeu/controller/LoginServlet.java [hunk 5]
```diff
@@ -59,28 +58,33 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
 
-    if (!username.matches("[\\w*\\s*]*")) {
-      request.setAttribute("error", "Please enter only letters, numbers, and spaces.");
-      request.getRequestDispatcher("/WEB-INF/view/login.jsp").forward(request, response);
-      return;
+    if (userStore.isUserRegistered(username)) {
+      User user = userStore.getUser(username);
+      if(password.equals(user.getPassword())){
+        request.getSession().setAttribute("user", username);
+	    response.sendRedirect("/conversations");
```
Comment (TylerSandman) [inline src/main/java/codeu/controller/LoginServlet.java]: Nit: Please fix indentation to be consistent

File: src/main/java/codeu/controller/LoginServlet.java [hunk 6]
```diff
@@ -59,28 +58,33 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
     String username = request.getParameter("username");
+    String password = request.getParameter("password");
 
-    if (!username.matches("[\\w*\\s*]*")) {
-      request.setAttribute("error", "Please enter only letters, numbers, and spaces.");
-      request.getRequestDispatcher("/WEB-INF/view/login.jsp").forward(request, response);
-      return;
+    if (userStore.isUserRegistered(username)) {
+      User user = userStore.getUser(username);
+      if(password.equals(user.getPassword())){
```
Comment (TylerSandman) [inline src/main/java/codeu/controller/LoginServlet.java]: Nit: Add space after "if"

if(password..... ) ---> if (password......)

File: src/main/java/codeu/controller/LoginServlet.java [hunk 7]
```diff
@@ -59,28 +58,33 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
+	    request.setAttribute("error", "Invalid password.");
+	    request.getRequestDispatcher("/WEB-INF/view/login.jsp").forward(request, response);
+      }
     }
 
-    if (!userStore.isUserRegistered(username)) {
-      //TODO: add password
-      User user = new User(UUID.randomUUID(), username, null, Instant.now());
-      userStore.addUser(user);
+    else{
```
Comment (TylerSandman) [inline src/main/java/codeu/controller/LoginServlet.java]: Nit: Add a space after else

else{ ----> else {

File: src/main/java/codeu/controller/LoginServlet.java [hunk 8]
```diff
@@ -59,28 +58,33 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
+	  request.setAttribute("error", "That username was not found.");
```
Comment (TylerSandman) [inline src/main/java/codeu/controller/LoginServlet.java]: Nit: Please keep indentation consistent and use 2 spaces

File: src/main/java/codeu/controller/LoginServlet.java [hunk 9]
```diff
@@ -59,28 +58,33 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
-      request.setAttribute("error", "Please enter only letters, numbers, and spaces.");
-      request.getRequestDispatcher("/WEB-INF/view/login.jsp").forward(request, response);
-      return;
+    if (userStore.isUserRegistered(username)) {
+      User user = userStore.getUser(username);
+      if(password.equals(user.getPassword())){
+        request.getSession().setAttribute("user", username);
+	    response.sendRedirect("/conversations");
+      }else{
+	    request.setAttribute("error", "Invalid password.");
```
Comment (TylerSandman) [inline src/main/java/codeu/controller/LoginServlet.java]: Nit: Please keep indentation consistent and use 2 spaces

File: src/main/java/codeu/controller/LoginServlet.java (hunk 6 above)
Comment (TylerSandman) [inline src/main/java/codeu/controller/LoginServlet.java]: Nit: Add space before curly bracket

if(password.equals(user.getPassword())){ --> if(password.equals(user.getPassword())) {

**Review by TylerSandman — COMMENTED** [review]
The code itself looks good but please pay attention to style, namely indentation and spacing. It's important that our code is clean and consistent, especially in a shared code base. Once fixed feel free to merge : )

**Review by CalamariDude — APPROVED** [review]
LGTM, tests passing, I should have seen this earlier, I apologize.

**Comment by nl356** [comment]
Looks good to me. I think it's safe to merge now @trishazalani

**Pull request merged**